js-packages/@prestojs/rest/src/Endpoint.ts

```diff
@@ -326,32 +355,71 @@ export default class Endpoint {
      * and `Endpoint.defaultConfig.requestInit`. Options passed here will take precedence. Only the `headers` key will be merged if it
      * exists in multiple places (eg. defaultConfig may include headers you want on every request). If you need to remove
      * a header entirely set the value to `undefined`.
      */
-    async execute({ urlArgs = {}, query, ...init }: PrepareOptions = {}): Promise<any> {
+    async execute(options: EndpointExecuteOptions = {}): Promise<ExecuteReturnVal> {
+        if (options.paginator) {
+            options = options.paginator.getRequestInit(options);
+        }
+        const { urlArgs = {}, query, paginator, ...init } = options;
         const url = this.urlPattern.resolve(urlArgs, { query });
         try {
             const cls = Object.getPrototypeOf(this).constructor;
-            return fetch(
+            const requestInit = mergeRequestInit(
+                cls.defaultConfig.requestInit,
+                this.requestInit,
+                init
+            );
+            const returnVal: ExecuteReturnVal = {
                 url,
-                mergeRequestInit(cls.defaultConfig.requestInit, this.requestInit, init)
-            )
+                urlArgs,
+                query,
+                requestInit,
+                result: null,
+            };
+            returnVal.result = await fetch(url, requestInit)
                 .then(async res => {
+                    returnVal.response = res;
                     if (!res.ok) {
                         throw new ApiError(res.status, res.statusText, await this.decodeBody(res));
                     }
                     return res;
                 })
                 .then(res => this.decodeBody(res))
-                .then(data =>
-                    this.transformResponseBody ? this.transformResponseBody(data) : data
-                );
+                .then(data => {
+                    returnVal.decodedBody = data;
+                    if (paginator) {
+                        // If we have a paginator update it's state based on response
+                        // This runs for all requests but should return false if response
+                        // is not paginated.
+                        const paginationState = cls.defaultConfig.getPaginationState(
+                            paginator,
+                            returnVal
+                        );
+                        if (paginationState) {
+                            paginator.setResponse(paginationState);
+                            data = paginationState.results;
+                        } else {
+                            // TODO: If you specify a paginator and the response is not paginated should
+                            // we warn? I think yes as I can't think of a reason why not but may need to
+                            // revisit this if we find a usecase
```

Comment:
it would be empty in that case

this is for where the response itself is definitely not a paginated shape

I _think_ we will probably want to remove this but I'll leave it for now until the case comes up. I suspect we'll want to write a hook or generic UI component that won't know upfront and so will create a paginator in case it's required